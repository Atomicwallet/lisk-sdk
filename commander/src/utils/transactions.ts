/*
 * LiskHQ/lisk-commander
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import {
	BaseTransaction,
	DelegateTransaction,
	MultisignatureTransaction,
	TransactionJSON,
	TransferTransaction,
	VoteTransaction,
} from '@liskhq/lisk-transactions';

import { ValidationError } from './error';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const parseTransactionString = (transactionStr: string): any => {
	try {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
		return JSON.parse(transactionStr);
	} catch (error) {
		throw new ValidationError('Could not parse transaction JSON.');
	}
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const defaultTransactions: { readonly [key: number]: any } = {
	8: TransferTransaction,
	10: DelegateTransaction,
	11: VoteTransaction,
	12: MultisignatureTransaction,
};

export const instantiateTransaction = (data: TransactionJSON): BaseTransaction => {
	if (data.type === undefined || data.type === null) {
		throw new Error('Invalid transaction without type');
	}

	if (!Object.keys(defaultTransactions).includes(String(data.type))) {
		throw new Error(`Transaction type ${data.type} is not supported`);
	}
	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
	const TransactionClass = defaultTransactions[data.type];

	// eslint-disable-next-line @typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call
	return new TransactionClass(data);
};
