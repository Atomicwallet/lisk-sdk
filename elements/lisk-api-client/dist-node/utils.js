"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toQueryString = function (obj) {
    var parts = Object.keys(obj).reduce(function (accumulator, key) { return __spread(accumulator, [
        encodeURIComponent(key) + "=" + encodeURIComponent(obj[key]),
    ]); }, []);
    return parts.join('&');
};
var urlParamRegex = /{[^}]+}/;
exports.solveURLParams = function (url, params) {
    if (params === void 0) { params = {}; }
    if (Object.keys(params).length === 0) {
        if (url.match(urlParamRegex) !== null) {
            throw new Error('URL is not completely solved');
        }
        return url;
    }
    var solvedURL = Object.keys(params).reduce(function (accumulator, key) {
        return accumulator.replace("{" + key + "}", params[key]);
    }, url);
    if (solvedURL.match(urlParamRegex) !== null) {
        throw new Error('URL is not completely solved');
    }
    return encodeURI(solvedURL);
};
//# sourceMappingURL=utils.js.map