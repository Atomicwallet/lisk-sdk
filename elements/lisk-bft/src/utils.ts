/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

import { validator } from '@liskhq/lisk-validator';

import { blockHeaderSchema } from './schema';
import { BlockHeader } from './types';

export const validateBlockHeader = (blockHeader: BlockHeader): boolean => {
	const errors = validator.validate(blockHeaderSchema, blockHeader);
	if (errors.length) {
		throw new Error(errors[0].message);
	}

	return true;
};
