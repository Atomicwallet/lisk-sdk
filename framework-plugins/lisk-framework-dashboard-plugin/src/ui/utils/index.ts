/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { apiClient } from '@liskhq/lisk-client';
import { Block, Transaction } from '../types';

const MAX_BLOCKS = 103 * 3;
const MAX_TRANSACTIONS = 150;

export const updateStatesOnNewBlock = (
	client: apiClient.APIClient,
	newBlockStr: string,
	blocks: Block[],
	confirmedTransactions: Transaction[],
	unconfirmedTransactions: Transaction[],
): {
	blocks: Block[];
	confirmedTransactions: Transaction[];
	unconfirmedTransactions: Transaction[];
} => {
	const newBlock = client.block.toJSON(client.block.decode(newBlockStr));
	const newBlocks = [...blocks, newBlock].slice(-1 * MAX_BLOCKS) as Block[];

	for (const trs of ((newBlock as unknown) as Block).payload) {
		confirmedTransactions.push(trs);
	}
	const confirmedTransactionsIds = confirmedTransactions.map(t => t.id);
	const newUnconfirmedTransactions = unconfirmedTransactions
		.filter(t => !confirmedTransactionsIds.includes(t.id))
		.slice(-1 * MAX_TRANSACTIONS);

	const newConfirmedTransactions = confirmedTransactions.slice(-1 * MAX_TRANSACTIONS);

	return {
		blocks: newBlocks,
		confirmedTransactions: newConfirmedTransactions,
		unconfirmedTransactions: newUnconfirmedTransactions,
	};
};

export const updateStatesOnNewTransaction = (
	client: apiClient.APIClient,
	newTransactionStr: string,
	unconfirmedTransactions: Transaction[],
): Transaction[] => {
	const transaction = client.transaction.toJSON(client.transaction.decode(newTransactionStr));
	return [...unconfirmedTransactions, transaction].slice(-1 * MAX_TRANSACTIONS) as Transaction[];
};
