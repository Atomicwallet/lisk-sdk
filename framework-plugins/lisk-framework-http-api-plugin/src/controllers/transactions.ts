/*
 * Copyright © 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { isBase64String } from '@liskhq/lisk-validator';
import { Request, Response } from 'express';
import { BaseChannel, PluginCodec } from 'lisk-framework';

export const getTransaction = (
	channel: BaseChannel,
	codec: PluginCodec,
) => async (req: Request, res: Response): Promise<void> => {
	const transactionId = req.params.id;

	// 400 - Client Side Error
	if (!transactionId || !isBase64String(transactionId)) {
		res.status(400).send({
			errors: [
				{ message: 'The transaction id parameter should be a base64 string.' },
			],
		});
		return;
	}

	let transaction: string;

	try {
		transaction = await channel.invoke<string>('app:getTransactionByID', {
			id: transactionId,
		});
	} catch (error) {
		if (
			(error as Error).message ===
			'Specified key transactions:id:×m does not exist'
		) {
			// 404 - Not Found Error
			res.status(404).send({
				errors: [
					{ message: `The transaction with id ${transactionId} not found.` },
				],
			});
			return;
		}

		// 500 - Server Error
		throw error;
	}

	// 200 - Response
	res.status(200).json(codec.decodeTransaction(transaction));
};
