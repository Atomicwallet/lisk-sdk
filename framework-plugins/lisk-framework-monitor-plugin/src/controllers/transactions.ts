/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Request, Response } from 'express';
import { BaseChannel } from 'lisk-framework';
import { PeerInfo, SharedState } from '../types';

const getAverage = (
	transactionsStats: Pick<SharedState['transactions'], 'transactions'>,
): number => {
	let transactionCount = 0;
	let total = 0;

	for (const transactionStats of Object.values(transactionsStats.transactions)) {
		transactionCount += 1;
		total += transactionStats.count;
	}

	return total / transactionCount;
};

export const getTransactionStats = (channel: BaseChannel, state: SharedState) => async (
	_req: Request,
	res: Response,
): Promise<void> => {
	const { transactions } = state;
	res.json({
		data: {
			transactions: transactions.transactions,
			connectedPeers: (await channel.invoke<ReadonlyArray<PeerInfo>>('app:getConnectedPeers'))
				.length,
			averageReceivedTransactions: getAverage(transactions),
		},
		meta: {},
	});
};
