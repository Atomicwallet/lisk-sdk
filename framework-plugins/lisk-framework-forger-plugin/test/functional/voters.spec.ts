/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Application } from 'lisk-framework';
import axios from 'axios';
import {
	createApplication,
	closeApplication,
	getURL,
	callNetwork,
	waitNBlocks,
	waitTill,
} from '../utils/application';
import { createVoteTransaction } from '../utils/transactions';

describe('Voters endpoint', () => {
	let app: Application;
	let accountNonce = 0;
	let networkIdentifier: Buffer;

	beforeAll(async () => {
		app = await createApplication('forger_functional_voters');
		// The test application generates a dynamic genesis block so we need to get the networkID like this
		networkIdentifier = app['_node'].networkIdentifier;
	});

	afterAll(async () => {
		await closeApplication(app);
	});

	describe('GET /api/voters/', () => {
		describe('200', () => {
			it('should return valid response', async () => {
				// Arrange & Act
				const { response, status } = await callNetwork(axios.get(getURL('/api/voters')));

				// Assert
				expect(status).toEqual(200);
				expect(response).toMatchSnapshot();
			});

			it('should return valid format', async () => {
				// Arrange & Act
				const { response, status } = await callNetwork(axios.get(getURL('/api/voters')));
				const forgerInfo = response.data[0];

				// Assert
				expect(status).toEqual(200);
				expect(response).toHaveProperty('data');
				expect(response).toHaveProperty('meta');
				expect(response.data).toBeInstanceOf(Array);
				expect(forgerInfo).toMatchObject(
					expect.objectContaining({
						address: expect.any(String),
						username: expect.any(String),
						totalVotesReceived: expect.any(String),
						voters: expect.any(Array),
					}),
				);
			});

			it('should return valid voters', async () => {
				// Arrange
				const { response: delegateResponse } = await callNetwork(axios.get(getURL('/api/voters')));
				const forgingDelegateAddress = delegateResponse.data[0].address;
				const transaction = createVoteTransaction({
					amount: '10',
					recipientAddress: forgingDelegateAddress,
					fee: '0.3',
					nonce: accountNonce,
					networkIdentifier,
				});
				accountNonce += 1;

				await app['_channel'].invoke('app:postTransaction', {
					transaction: transaction.getBytes().toString('hex'),
				});
				await waitNBlocks(app, 1);
				// Wait a bit to give plugin a time to calculate forger info
				await waitTill(2000);

				// Act
				const { response, status } = await callNetwork(axios.get(getURL('/api/voters')));
				const forgerInfo = response.data.find(
					(forger: any) => forger.address === forgingDelegateAddress,
				);

				// Assert
				expect(status).toEqual(200);
				expect(forgerInfo.voters[0]).toMatchObject(
					expect.objectContaining({
						address: transaction.senderID.toString('hex'),
						amount: '1000000000',
					}),
				);
			});
		});
	});
});
