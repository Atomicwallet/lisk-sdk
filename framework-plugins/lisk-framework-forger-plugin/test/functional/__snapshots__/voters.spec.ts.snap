// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Voters endpoint GET /api/voters/ 200 should return valid response 1`] = `
Object {
  "data": Array [
    Object {
      "address": "nKvuPSdCZna4Us5rgEyy/f980LU=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_1",
      "voters": Array [],
    },
    Object {
      "address": "Rj5+h5t73GqX7AKipgOqGkagTIA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_2",
      "voters": Array [],
    },
    Object {
      "address": "Mo0PVGaVxfoCEF3rBVzygB2bi6E=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_3",
      "voters": Array [],
    },
    Object {
      "address": "YXRRX6Zskb/xEokT7dTg8d43zuA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_4",
      "voters": Array [],
    },
    Object {
      "address": "nQFJsJYtRL/Aip9k1a/Otigdf7U=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_5",
      "voters": Array [],
    },
    Object {
      "address": "tWxVuacMji8Hl5uGI3Su0Okqbdo=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_6",
      "voters": Array [],
    },
    Object {
      "address": "t1gJad1WFR9giTHxJveTu/Rdj6A=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_7",
      "voters": Array [],
    },
    Object {
      "address": "WNkH0mUIYD6DhCPaogYcKceoSVA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_8",
      "voters": Array [],
    },
    Object {
      "address": "UvnNz/BgUkHHgnhpCuNusBNqMP8=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_9",
      "voters": Array [],
    },
    Object {
      "address": "tIW+zYjbGrPVVtQFIERRugCtqn0=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_10",
      "voters": Array [],
    },
    Object {
      "address": "T9UvZ/FR+73andkqcUiEo5mDDso=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_11",
      "voters": Array [],
    },
    Object {
      "address": "JMEw62zA2PZjqPbRb/xh+TWi4C4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_12",
      "voters": Array [],
    },
    Object {
      "address": "45MWzAIAiep6VhS89pqJMcEGMKc=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_13",
      "voters": Array [],
    },
    Object {
      "address": "a5iVwxvNstnJKbnafjie2R3mcqA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_14",
      "voters": Array [],
    },
    Object {
      "address": "Gsc7/3SSStm3QjbElivicXSuh9A=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_15",
      "voters": Array [],
    },
    Object {
      "address": "Mf54m0MnfjWrQQ8q/PtXQoCvLdg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_16",
      "voters": Array [],
    },
    Object {
      "address": "oGIEcs3gPnfK7OcBq3vFkopdNnw=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_17",
      "voters": Array [],
    },
    Object {
      "address": "9zDLkpocRQMjh8NF4Q0kJ76lWl4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_18",
      "voters": Array [],
    },
    Object {
      "address": "pvagVDrkcMawVgIcsqwVM2jq/uw=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_19",
      "voters": Array [],
    },
    Object {
      "address": "efMMHLwbnElJyLhazCSnV44BVYs=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_20",
      "voters": Array [],
    },
    Object {
      "address": "y1ee5TezSSbUcSmgtUwObQDvMAQ=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_21",
      "voters": Array [],
    },
    Object {
      "address": "Q2tA9YwMJ+0TP2ABoBn/JVYe+tQ=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_22",
      "voters": Array [],
    },
    Object {
      "address": "TodLz7b1iW/p5dqzsm9ZsuKpwJs=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_23",
      "voters": Array [],
    },
    Object {
      "address": "0smpN1Wu0gxNj1XB6SuBLSx9SdI=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_24",
      "voters": Array [],
    },
    Object {
      "address": "PeleGPGKVOImm7+PGjjqcHYsc/o=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_25",
      "voters": Array [],
    },
    Object {
      "address": "3w4Ye7OJWAYmHIfPZuF3JWbujlg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_26",
      "voters": Array [],
    },
    Object {
      "address": "von06YPfsE4rWKEuue0YFJ4Qiwc=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_27",
      "voters": Array [],
    },
    Object {
      "address": "S2EmWXiBy2uhpFwfYoZ2nnoJT7Q=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_28",
      "voters": Array [],
    },
    Object {
      "address": "cKvwVr2S6Pd8/FUXSPpUpOMBjV8=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_29",
      "voters": Array [],
    },
    Object {
      "address": "0KDkW5UOOHHYeDuXNAkEK0qzgtQ=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_30",
      "voters": Array [],
    },
    Object {
      "address": "Pe6wp0JqAotDW03djTWshc9Wcjc=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_31",
      "voters": Array [],
    },
    Object {
      "address": "X71EKkZHsHnNoSKez22PRPNhyMo=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_32",
      "voters": Array [],
    },
    Object {
      "address": "rUL46GfWGBcb9JguZCaUQhSPbhE=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_33",
      "voters": Array [],
    },
    Object {
      "address": "A/bZC329BJfcOlLRwn4ju4x1iX8=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_34",
      "voters": Array [],
    },
    Object {
      "address": "kTnJH4oKofs4V3D+rymbmYg67C0=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_35",
      "voters": Array [],
    },
    Object {
      "address": "YfOW0qShOrejm6eR+sS5IbVKII4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_36",
      "voters": Array [],
    },
    Object {
      "address": "hyJFM4P3gdVCek7iEQIOSb80ork=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_37",
      "voters": Array [],
    },
    Object {
      "address": "gHTw0C90j8VUSKS/IA8drehRcFk=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_38",
      "voters": Array [],
    },
    Object {
      "address": "MIqV0dP3u1VvSNpPQ0RWbln28cs=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_39",
      "voters": Array [],
    },
    Object {
      "address": "1b0gULdLMJ1UgZyhet0XPG/KHhY=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_40",
      "voters": Array [],
    },
    Object {
      "address": "ixwiGgMM9yBzbZ+30Emd1ydvwbM=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_41",
      "voters": Array [],
    },
    Object {
      "address": "YzD9iukd9KXX+8I5DBguxmdtxaY=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_42",
      "voters": Array [],
    },
    Object {
      "address": "lBRsmIl0jHtyfrOsjCDlPFLv/TI=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_43",
      "voters": Array [],
    },
    Object {
      "address": "KQq8SiJEvw7PWqHM7orI9g+Lzkg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_44",
      "voters": Array [],
    },
    Object {
      "address": "q9LtWtNbOghwqtrm3OrMmIumOJU=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_45",
      "voters": Array [],
    },
    Object {
      "address": "HBlMK+HMU/ZjqTxkiZy6o0AW9BU=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_46",
      "voters": Array [],
    },
    Object {
      "address": "XNHQzPmPK9Wkv6p3DVXxZJivC8w=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_47",
      "voters": Array [],
    },
    Object {
      "address": "ZX9hByju+X1V5QIShx8Jk7t8xwA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_48",
      "voters": Array [],
    },
    Object {
      "address": "IVn3XlRAw2QxrtvH3CmmWjJ3eLg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_49",
      "voters": Array [],
    },
    Object {
      "address": "DzOlAzt1DmxNykfji6Ag6RLfFD4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_50",
      "voters": Array [],
    },
    Object {
      "address": "m0LkJkAg88Pcqu2AZXjM1GkgUGA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_51",
      "voters": Array [],
    },
    Object {
      "address": "rr2Z8HIYEJFiqQXQ4MkeWL7cg8U=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_52",
      "voters": Array [],
    },
    Object {
      "address": "gsvHs501rzWPniUTrxOy93tkegA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_53",
      "voters": Array [],
    },
    Object {
      "address": "DSw3fpNraMcAZmE7EMD9rVN/kNo=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_54",
      "voters": Array [],
    },
    Object {
      "address": "ibFE7P3V6jUgg79iTTz4QuwGpeM=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_55",
      "voters": Array [],
    },
    Object {
      "address": "LPUsCMx2CR2IToAMHGl7E/aZB9Q=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_56",
      "voters": Array [],
    },
    Object {
      "address": "4pUKnwe0TnJN8hKTYMwUApPAgwg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_57",
      "voters": Array [],
    },
    Object {
      "address": "ZfknGHv5aqxdlo/Mk1HlSStfk1Y=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_58",
      "voters": Array [],
    },
    Object {
      "address": "0G/m0+X3+stYVeyoOUIv44JKXW4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_59",
      "voters": Array [],
    },
    Object {
      "address": "bhLkSYrmn7B/8tiqsDapESKdbGI=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_60",
      "voters": Array [],
    },
    Object {
      "address": "3LW/NbbVIRleYTxCSD9SATniMx0=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_61",
      "voters": Array [],
    },
    Object {
      "address": "hQbzwQ91BElG8aI6fK9XglNklHE=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_62",
      "voters": Array [],
    },
    Object {
      "address": "08gGTQEe+FPjvlBrlaBF9B945yo=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_63",
      "voters": Array [],
    },
    Object {
      "address": "yYVUEjBirFeVo+6QWwgehj21qBg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_64",
      "voters": Array [],
    },
    Object {
      "address": "oLxQsn56w5Bg7QFaVfL0UIyE8MI=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_65",
      "voters": Array [],
    },
    Object {
      "address": "1cTjgLHsL38gaM+6mpDLOueBYRA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_66",
      "voters": Array [],
    },
    Object {
      "address": "fSxngdhz7SunqH9G9zX14VpBpvE=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_67",
      "voters": Array [],
    },
    Object {
      "address": "Oz4TexvsbyDJqLKtT1eEZh+w+nk=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_68",
      "voters": Array [],
    },
    Object {
      "address": "b/zYrVR9ilSaMbJSNuMix4GlLYU=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_69",
      "voters": Array [],
    },
    Object {
      "address": "vRdXKdQXclnHHPE/1Oz7XQFUJwY=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_70",
      "voters": Array [],
    },
    Object {
      "address": "WFOj8kmQ3uzO1J1rwVmQEC7Awzo=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_71",
      "voters": Array [],
    },
    Object {
      "address": "6TVRUsEXyeH62L6G6avqlhzvSjY=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_72",
      "voters": Array [],
    },
    Object {
      "address": "T0Qi62HEXttNdvEM2HHJ+YPy66o=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_73",
      "voters": Array [],
    },
    Object {
      "address": "CQP0xctZmnkorvJ+MU6YKR0eOIg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_74",
      "voters": Array [],
    },
    Object {
      "address": "MSBK1bld2SLCiZqlv45+5bdUavM=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_75",
      "voters": Array [],
    },
    Object {
      "address": "hFm4hw/O//WfFy1xa3v+n8wo1Ag=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_76",
      "voters": Array [],
    },
    Object {
      "address": "JG+6XFGVdtk8X6yJnESym3L1Jq4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_77",
      "voters": Array [],
    },
    Object {
      "address": "2OYRuv1wpUnwNc9hqw1u2efyXE4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_78",
      "voters": Array [],
    },
    Object {
      "address": "fWDbGHM3y9iBFA1p2EySRu2oOC4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_79",
      "voters": Array [],
    },
    Object {
      "address": "sRxYEeoHSjAULYJLboz9PfFLJog=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_80",
      "voters": Array [],
    },
    Object {
      "address": "js7/1aQeZ4tkZ8m8gM410uhUPZg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_81",
      "voters": Array [],
    },
    Object {
      "address": "yl9tduq25PWqzuKGTHkDTXERuYY=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_82",
      "voters": Array [],
    },
    Object {
      "address": "C8O+wv21ZZlv0xbjaOZuXY6DCAg=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_83",
      "voters": Array [],
    },
    Object {
      "address": "qwBBp9P3ssKQtbg01Gvce364WBU=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_84",
      "voters": Array [],
    },
    Object {
      "address": "xpe2IMfEAV4y3XvdfQQwszQE4Qc=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_85",
      "voters": Array [],
    },
    Object {
      "address": "w6sqwjUS2b9isCd14iz4DfgU6xs=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_86",
      "voters": Array [],
    },
    Object {
      "address": "yjCaX0u/EcqGWS/rttLMx4MJ9p4=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_87",
      "voters": Array [],
    },
    Object {
      "address": "J4Q6YKHgRMHm488Rn99k6ys+DZQ=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_88",
      "voters": Array [],
    },
    Object {
      "address": "isgAEk1bFq/Ve1z3JF7fzViF6js=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_89",
      "voters": Array [],
    },
    Object {
      "address": "oo1eNAB/2P5teQMETrI6YP2tPAA=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_90",
      "voters": Array [],
    },
    Object {
      "address": "1eH1LL5KEaNzC5j1IQm1dgKpxKE=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_91",
      "voters": Array [],
    },
    Object {
      "address": "PIDn2ZZKHIOm3V3GThBeDmNL1Yo=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_92",
      "voters": Array [],
    },
    Object {
      "address": "+lJqFhHMxm3sgVy5YxdBGAdLc24=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_93",
      "voters": Array [],
    },
    Object {
      "address": "O5bYVlVpQh9DaEssTqoGOcu14BE=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_94",
      "voters": Array [],
    },
    Object {
      "address": "rPvbrrk9WHFwx82cC1/963/52uw=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_95",
      "voters": Array [],
    },
    Object {
      "address": "/86M4iXF2ACY9Q6HcSW2Va720QE=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_96",
      "voters": Array [],
    },
    Object {
      "address": "qcZmlN1lsv30DN9FoMMIy9OABPw=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_97",
      "voters": Array [],
    },
    Object {
      "address": "CtpqL2yPiRdpNm/Jqm/Z8frLNs8=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_98",
      "voters": Array [],
    },
    Object {
      "address": "t2oPGBnEvgoUglZ8qbn77T7aREw=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_99",
      "voters": Array [],
    },
    Object {
      "address": "tUPi5ZIgC+s4I19uSPir4dh62HI=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_100",
      "voters": Array [],
    },
    Object {
      "address": "T9jMTiejSJtX7Zhu/j0yfT3kDZI=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_101",
      "voters": Array [],
    },
    Object {
      "address": "k286D013a2p3Iu0Sbo/xe0TX57g=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_102",
      "voters": Array [],
    },
    Object {
      "address": "Wt5WQ5nmcL0dQpWDBZBn86bKK38=",
      "totalVotesReceived": "1000000000000",
      "username": "genesis_103",
      "voters": Array [],
    },
  ],
  "meta": Object {
    "count": 103,
  },
}
`;
